File: src/ormai/core/context.py
"""
Execution context for OrmAI tool calls.
"""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any
from uuid import uuid4


@dataclass(frozen=True)
class Principal:
    """
    Represents the authenticated user/tenant making a request.

    This is the identity that policies are evaluated against for scoping,
    ACL checks, and audit logging.
    """

    tenant_id: str
    user_id: str
    roles: tuple[str, ...] = field(default_factory=tuple)
    metadata: dict[str, Any] = field(default_factory=dict)

    def has_role(self, role: str) -> bool:
        """Check if the principal has a specific role."""
        return role in self.roles

    def has_any_role(self, *roles: str) -> bool:
        """Check if the principal has any of the specified roles."""
        return bool(set(roles) & set(self.roles))


@dataclass
class RunContext:
    """
    Execution context for a single tool call.

    Carries all the information needed to execute a tool call safely:
    - Principal identity for scoping and ACLs
    - Request tracking for auditing
    - Database session for the operation
    - Timing information
    """

    principal: Principal
    db: Any  # Database session (SQLAlchemy Session, Tortoise connection, etc.)
    request_id: str = field(default_factory=lambda: str(uuid4()))
    trace_id: str | None = None
    now: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    metadata: dict[str, Any] = field(default_factory=dict)

    @classmethod
    def create(
        cls,
        *,
        tenant_id: str,
        user_id: str,
        db: Any,
        roles: tuple[str, ...] | list[str] = (),
        request_id: str | None = None,
        trace_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> "RunContext":
        """
        Convenience factory for creating a RunContext.

        Args:
            tenant_id: The tenant ID for scoping
            user_id: The user ID for auditing
            db: The database session
            roles: User roles for ACL checks
            request_id: Optional request ID (generated if not provided)
            trace_id: Optional trace ID for distributed tracing
            metadata: Optional additional metadata
        """
        principal = Principal(
            tenant_id=tenant_id,
            user_id=user_id,
            roles=tuple(roles) if isinstance(roles, list) else roles,
        )
        return cls(
            principal=principal,
            db=db,
            request_id=request_id or str(uuid4()),
            trace_id=trace_id,
            metadata=metadata or {},
        )

File: ormai-ts/src/core/context.ts
/**
 * Execution context for OrmAI tool calls.
 *
 * The context carries identity information, database connection, and request
 * metadata through the entire tool execution pipeline.
 */

import { randomUUID } from 'crypto';

/**
 * Identity information for the executing user/tenant.
 *
 * Principal is immutable and carries all identity claims needed for
 * policy evaluation and audit logging.
 */
export interface Principal {
  /** Tenant identifier for multi-tenancy scoping */
  readonly tenantId: string;

  /** User identifier for audit logging */
  readonly userId: string;

  /** Roles for role-based access control */
  readonly roles: readonly string[];

  /** Additional identity metadata */
  readonly metadata?: Readonly<Record<string, unknown>>;
}

/**
 * Create a new Principal.
 */
export function createPrincipal(opts: {
  tenantId: string;
  userId: string;
  roles?: readonly string[];
  metadata?: Record<string, unknown>;
}): Principal {
  return Object.freeze({
    tenantId: opts.tenantId,
    userId: opts.userId,
    roles: Object.freeze(opts.roles ?? []),
    metadata: opts.metadata ? Object.freeze({ ...opts.metadata }) : undefined,
  });
}

/**
 * Check if a principal has a specific role.
 */
export function hasRole(principal: Principal, role: string): boolean {
  return principal.roles.includes(role);
}

/**
 * Check if a principal has any of the specified roles.
 */
export function hasAnyRole(principal: Principal, ...roles: string[]): boolean {
  return roles.some((role) => principal.roles.includes(role));
}

/**
 * Execution context for a tool call.
 *
 * RunContext is created for each request and carries all information
 * needed for policy evaluation, query execution, and auditing.
 *
 * @typeParam DB - The database client type (e.g., PrismaClient)
 */
export interface RunContext<DB = unknown> {
  /** Identity of the executing user/tenant */
  readonly principal: Principal;

  /** Database client/connection */
  readonly db: DB;

  /** Unique request identifier for tracing */
  readonly requestId: string;

  /** Distributed trace ID (optional) */
  readonly traceId?: string;

  /** Current timestamp at request start */
  readonly now: Date;

  /** Additional request metadata */
  readonly metadata?: Readonly<Record<string, unknown>>;
}

/**
 * Options for creating a RunContext.
 */
export interface CreateContextOptions<DB> {
  /** Tenant ID */
  tenantId: string;

  /** User ID */
  userId: string;

  /** Database client */
  db: DB;

  /** User roles (optional) */
  roles?: readonly string[];

  /** Request ID (auto-generated if not provided) */
  requestId?: string;

  /** Trace ID for distributed tracing */
  traceId?: string;

  /** Additional metadata */
  metadata?: Record<string, unknown>;
}

/**
 * Create a new RunContext.
 */
export function createContext<DB>(opts: CreateContextOptions<DB>): RunContext<DB> {
  const principal = createPrincipal({
    tenantId: opts.tenantId,
    userId: opts.userId,
    roles: opts.roles,
  });

  return Object.freeze({
    principal,
    db: opts.db,
    requestId: opts.requestId ?? randomUUID(),
    traceId: opts.traceId,
    now: new Date(),
    metadata: opts.metadata ? Object.freeze({ ...opts.metadata }) : undefined,
  });
}

/**
 * Create a context with full options including a pre-built Principal.
 */
export function createContextWithPrincipal<DB>(opts: {
  principal: Principal;
  db: DB;
  requestId?: string;
  traceId?: string;
  metadata?: Record<string, unknown>;
}): RunContext<DB> {
  return Object.freeze({
    principal: opts.principal,
    db: opts.db,
    requestId: opts.requestId ?? randomUUID(),
    traceId: opts.traceId,
    now: new Date(),
    metadata: opts.metadata ? Object.freeze({ ...opts.metadata }) : undefined,
  });
}

/**
 * Type guard to check if a value is a valid Principal.
 */
export function isPrincipal(value: unknown): value is Principal {
  if (typeof value !== 'object' || value === null) {
    return false;
  }
  const obj = value as Record<string, unknown>;
  return (
    typeof obj['tenantId'] === 'string' &&
    typeof obj['userId'] === 'string' &&
    Array.isArray(obj['roles'])
  );
}

/**
 * Type guard to check if a value is a valid RunContext.
 */
export function isRunContext<DB>(value: unknown): value is RunContext<DB> {
  if (typeof value !== 'object' || value === null) {
    return false;
  }
  const obj = value as Record<string, unknown>;
  return (
    isPrincipal(obj['principal']) &&
    typeof obj['requestId'] === 'string' &&
    obj['now'] instanceof Date
  );
}
